src/orange/detail/BIHBuilder.hh

```diff
@@ -0,0 +1,86 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022-2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/detail/BIHBuilder.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <variant>
+#include <vector>
+
+#include "corecel/cont/Range.hh"
+#include "orange/BoundingBox.hh"
+#include "orange/detail/BIHData.hh"
+#include "orange/detail/BIHPartitioner.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Create a bounding interval hierarchy from supplied bounding boxes.
+ *
+ * This implementation matches the structure proposed in the original
+ * paper [1]. Partitioning is done on the basis of bounding box centers using
+ * the "longest dimension" heuristic. All leaf nodes contain either a single
+ * volume id, or multiple volume ids if the volumes have bounding boxes that
+ * share the same center.
+ *
+ * [1] C. Wachter, Carsten and A. Keller, "Instant Ray Tracing: The Bounding
+ * Interval Hierarchy" Eurographics Symposium on Rendering, 2006,
+ * doi:10.2312/EGWR/EGSR06/139-149}
+ */
+class BIHBuilder
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using VecBBox = std::vector<FastBBox>;
+    //!@}
+
+  public:
+    //! Default constructor
+    BIHBuilder() = default;
+
+    // Construct from a Storage object
+    explicit BIHBuilder(BIHStorage storage);
+
+    // Create BIH Nodes
+    BIHTree operator()(VecBBox bboxes);
+
+  private:
+    /// TYPES ///
+
+    using Real3 = Array<fast_real_type, 3>;
```

Comment:
Possibly override the `real_type` here too?
```suggestion
    using real_type = fast_real_type;
    using Real3 = Array<real_type, 3>;
```

Reply:
This starts to get messy because we are passing BoundingBox<double> to this class. I thought we decided to "draw the line" between the doubles and floats at the point of storage?

Reply:
I mean, I could read them from disk as doubles too

Reply:
Same comment as above